### PR TITLE
[HDRP][Fix] Case 1158661: Hardware mode for dynamic resolution breaks editor and game views

### PR DESCRIPTION
**Pull request by NetteVI**
### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ]  Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Fixes the visual artifacts and issues with hardware dynamic resolution, and re-enables it for Metal.

- Specifically, this PR adds an overload of RTHandleSystem::SetReferenceSize that takes in the size of the final full resolution (without dyn-res applied) so that when hardware dynamic resolution is enabled, rtHandleScale stores the final resolution / maxSize. This is necessary to upscale to the final resolution correctly during the final pass, instead of upscaling to the maxSize of the render texture which causes the scaling artifacts that were previously seen when resizing the game view.
- This PR also prevents the Lanczos upsampling shader function from dividing by 0, which previously caused numerous black line artifacts, especially noticeable when resizing.
image

- Finally, this also ensures that the RTHandleSystem always populates the m_AutoSizedRTsArray when setting a new hardware dynamic resolution state so that render textures are always recreated with the correct dynamic resolution state (this fixes case 1158661: https://fogbugz.unity3d.com/f/cases/1158661/ where the Scene View wasn't turning dynamic resolution off).
 
---
### Testing status

**NOTE**: There was an existing issue with DX12 that causes the editor to crash when hardware dynamic resolution is enabled. That's fixed in this trunk PR: https://ono.unity3d.com/unity/unity/pull-request/111190/_/graphics/dynamic-resolution/case-1158661 and tests should be run against the custom revision: adc71429b87e

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [x] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
- Other: 

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/HDRP%252Fhardware-dynamic-resolution

### CODE REVIEW

**Comment by FrancescoC-unity** [comment]
I pushed some small changes

**Review by FrancescoC-unity — CHANGES_REQUESTED** [review]
Actually scratch this, it is an issue that was there before too.

**Review by FrancescoC-unity — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/RTUpscale.hlsl [hunk 1]
```diff
@@ -82,9 +81,13 @@ CTYPE Lanczos(TEXTURE2D_X(_InputTexture), float2 inUV)
     float2 xMin4 = x - 4.0;
     float2 xMin5 = x - 5.0;
 
+    // By the definition of x, only xMin2 should possibly = 0
+    xMin2 = (xMin2 == 0) * epsilon + xMin2; // if (xMin2 == 0), then xMin2 = epsilon
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/RTUpscale.hlsl:85]: won't this be simpler (and more efficient) to always do:

float2 xMin2 = x - 2.0 + epsilon;

epsilon is negligible anyway.

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/RTUpscale.hlsl:85]: ok forget it, don't solve the issue

**Comment by FrancescoC-unity** [comment]
**NOTE** This PR introduces some issue with TAA (I assume wrong history scale)

**Pull request merged**